tests/UnitTests/CoreGraphics.drawing/CGContextDrawing_PatternTests.cpp

```diff
@@ -0,0 +1,178 @@
+    for (int i = 1; i < 5; i++) {
+        CGContextAddLineToPoint(context, r * sin(i * theta), r * cos(i * theta));
+    }
+    CGContextClosePath(context);
+    CGContextFillPath(context);
+}
+
+static void drawStencilBox(void* info, CGContextRef context) {
+    CGContextFillRect(context, CGRectMake(0, 0, 8, 8));
+}
+
+struct PatternCallback {
+    const char* name;
+    CGPatternDrawPatternCallback callback;
+};
+
+struct StencilColor {
+    const char* name;
+    float color[4];
+};
+
+PatternCallback stencilPatternCallback[] = { { "drawStencilStar", &drawStencilStar }, { "drawStencilBox", &drawStencilBox } };
+
+StencilColor stencilColors[] = { { "orange", { 1, 0.64, 0, 1 } }, { "gold", { 1, 0.84, 0, 1 } },    { "red", { 1, 0, 0, 1 } },
+                                 { "green", { 0, 1, 0, 1 } },     { "blue", { 0, 0, 1, 1 } },       { "cyan", { 0, 1, 1, 1 } },
+                                 { "teal", { 0, 0.5, 0.5, 1 } },  { "purple", { 0.5, 0, 0.5, 1 } }, { "Maroon", { 0.5, 0, 0, 1 } } };
+
+int patternsizes[] = { 1, 2, 4, 8, 13, 10, 16, 20, 32, 128, 241 };
+
+//<PatternCallback,StencilColor,pattern size>
+class CGPatternStencil : public WhiteBackgroundTest<>,
+                         public ::testing::WithParamInterface<::testing::tuple<PatternCallback, StencilColor, int>> {
+    CFStringRef CreateOutputFilename() {
+        PatternCallback stencil = ::testing::get<0>(GetParam());
+        StencilColor stencilColor = ::testing::get<1>(GetParam());
+        int size = ::testing::get<2>(GetParam());
+
+        return CFStringCreateWithFormat(nullptr,
+                                        nullptr,
+                                        CFSTR("TestImage.CGContext.CGPatternStencil.%s.%s.%d.png"),
+                                        stencil.name,
+                                        stencilColor.name,
+                                        size);
+    }
+};
+
+TEST_P(CGPatternStencil, DrawStencils) {
+    CGPatternDrawPatternCallback callback = ::testing::get<0>(GetParam()).callback;
+    const CGFloat* stencilColor = ::testing::get<1>(GetParam()).color;
+    int size = ::testing::get<2>(GetParam());
+
+    CGContextRef context = GetDrawingContext();
+    CGRect bounds = GetDrawingBounds();
+
+    static const CGPatternCallbacks callbacks = { 0, callback, NULL };
+
+    CGColorSpaceRef baseSpace = CGColorSpaceCreateDeviceRGB();
+    CGColorSpaceRef patternSpace = CGColorSpaceCreatePattern(baseSpace);
+    CGContextSetFillColorSpace(context, patternSpace);
+    CGColorSpaceRelease(patternSpace);
+    CGColorSpaceRelease(baseSpace);
+
+    CGPatternRef pattern = CGPatternCreate(
+        NULL, CGRectMake(0, 0, size, size), CGAffineTransformIdentity, size, size, kCGPatternTilingConstantSpacing, false, &callbacks);
+
+    CGContextSetFillPattern(context, pattern, stencilColor);
+    CGPatternRelease(pattern);
+    CGContextFillRect(context, bounds);
+}
+
+INSTANTIATE_TEST_CASE_P(CGPatternTests,
```

Comment:
Do we need to verify so many times that they are getting clipped (with different colors? ) 😸